internal/app/machined/pkg/adapters/network/nftables_rule.go

```diff
@@ -0,0 +1,551 @@
+//
+//nolint:revive,golint
+func NfTablesRule(r *network.NfTablesRule) nftablesRule {
+	return nftablesRule{
+		NfTablesRule: r,
+	}
+}
+
+type nftablesRule struct {
+	*network.NfTablesRule
+}
+
+// SetKind is the type of the nftables Set.
+type SetKind uint8
+
+// SetKind constants.
+const (
+	SetKindIPv4 SetKind = iota
+	SetKindIPv6
+	SetKindPort
+)
+
+// NfTablesSet is a compiled representation of the set.
+type NfTablesSet struct {
+	Kind      SetKind
+	Addresses []netipx.IPRange
+	Ports     [][2]uint16
+}
+
+// KeyType returns the type of the set.
+func (set NfTablesSet) KeyType() nftables.SetDatatype {
+	switch set.Kind {
+	case SetKindIPv4:
+		return nftables.TypeIPAddr
+	case SetKindIPv6:
+		return nftables.TypeIP6Addr
+	case SetKindPort:
+		return nftables.TypeInetService
+	default:
+		panic(fmt.Sprintf("unknown set kind: %d", set.Kind))
+	}
+}
+
+// SetElements returns the set elements.
+func (set NfTablesSet) SetElements() []nftables.SetElement {
+	switch set.Kind {
+	case SetKindIPv4, SetKindIPv6:
+		elements := make([]nftables.SetElement, 0, len(set.Addresses)*2)
+
+		for _, r := range set.Addresses {
+			fromBin, _ := r.From().MarshalBinary()    //nolint:errcheck // doesn't fail
+			toBin, _ := r.To().Next().MarshalBinary() //nolint:errcheck // doesn't fail
+
+			elements = append(elements,
+				nftables.SetElement{
+					Key:         fromBin,
+					IntervalEnd: false,
+				},
+				nftables.SetElement{
+					Key:         toBin,
+					IntervalEnd: true,
+				},
+			)
+		}
+
+		return elements
+	case SetKindPort:
+		elements := make([]nftables.SetElement, 0, len(set.Ports))
+
+		for _, p := range set.Ports {
+			from := binaryutil.BigEndian.PutUint16(p[0])
+			to := binaryutil.BigEndian.PutUint16(p[1])
+
+			elements = append(elements,
+				nftables.SetElement{
+					Key:         from,
+					IntervalEnd: false,
+				},
+				nftables.SetElement{
+					Key:         to,
+					IntervalEnd: true,
+				},
+			)
+		}
+
+		return elements
+	default:
+		panic(fmt.Sprintf("unknown set kind: %d", set.Kind))
+	}
+}
+
+// NfTablesCompiled is a compiled representation of the rule.
+type NfTablesCompiled struct {
+	Rules [][]expr.Any
+	Sets  []NfTablesSet
+}
+
+var (
+	matchV4 = []expr.Any{
+		// Store protocol type to register 1
+		&expr.Meta{
+			Key:      expr.MetaKeyNFPROTO,
+			Register: 1,
+		},
+		// Match IP Family
+		&expr.Cmp{
+			Op:       expr.CmpOpEq,
+			Register: 1,
+			Data:     []byte{byte(nftables.TableFamilyIPv4)},
+		},
+	}
+
+	matchV6 = []expr.Any{
+		// Store protocol type to register 1
+		&expr.Meta{
+			Key:      expr.MetaKeyNFPROTO,
+			Register: 1,
+		},
+		// Match IP Family
+		&expr.Cmp{
+			Op:       expr.CmpOpEq,
+			Register: 1,
+			Data:     []byte{byte(nftables.TableFamilyIPv6)},
+		},
+	}
+)
+
+// Compile translates the rule into the set of nftables instructions.
+//
+//nolint:gocyclo,cyclop
+func (a nftablesRule) Compile() (*NfTablesCompiled, error) {
```

Comment:
this is the core compiler of the resource spec for ntftables rule down to nftables instructions